File: src/services/medicalCertificationService.ts
import {
  collection,
  doc,
  getDocs,
  updateDoc,
  deleteDoc,
  query,
  orderBy,
  Timestamp,
  onSnapshot,
  Unsubscribe
} from 'firebase/firestore';
import { ref, deleteObject } from 'firebase/storage';
import { logger } from '@/utils/logger';
import { db, storage } from '@/lib/firebase';
import { MobileMedicalCertificate, CertificateStatus } from '@/types/user.types';

/**
 * Service for managing medical certificates uploaded via the mobile app (CalyMob).
 * These certificates are stored in a subcollection under each member document.
 *
 * Firestore path: clubs/{clubId}/members/{memberId}/medical_certificates
 */
export class MedicalCertificationService {
  /**
   * Get all medical certificates for a member
   */
  static async getMemberCertifications(
    clubId: string,
    memberId: string
  ): Promise<MobileMedicalCertificate[]> {
    const certsRef = collection(db, `clubs/${clubId}/members/${memberId}/medical_certificates`);
    const q = query(certsRef, orderBy('uploaded_at', 'desc'));
    const snapshot = await getDocs(q);

    return snapshot.docs.map(doc => ({
      id: doc.id,
      ...doc.data()
    })) as MobileMedicalCertificate[];
  }

  /**
   * Watch medical certificates for a member (real-time updates)
   */
  static watchMemberCertifications(
    clubId: string,
    memberId: string,
    callback: (certs: MobileMedicalCertificate[]) => void
  ): Unsubscribe {
    const certsRef = collection(db, `clubs/${clubId}/members/${memberId}/medical_certificates`);
    const q = query(certsRef, orderBy('uploaded_at', 'desc'));

    return onSnapshot(q, (snapshot) => {
      const certs = snapshot.docs.map(doc => ({
        id: doc.id,
        ...doc.data()
      })) as MobileMedicalCertificate[];
      callback(certs);
    });
  }

  /**
   * Approve a medical certificate
   * Sets the status to 'approved', valid_until date, and reviewer info
   */
  static async approveCertification(
    clubId: string,
    memberId: string,
    certId: string,
    validUntil: Date,
    reviewerId: string,
    reviewerName: string
  ): Promise<void> {
    const certRef = doc(db, `clubs/${clubId}/members/${memberId}/medical_certificates/${certId}`);

    await updateDoc(certRef, {
      status: 'approved' as CertificateStatus,
      valid_until: Timestamp.fromDate(validUntil),
      reviewed_by: reviewerId,
      reviewed_by_nom: reviewerName,
      reviewed_at: Timestamp.fromDate(new Date())
    });

    // Also update the member's main validity date if this is newer
    const memberRef = doc(db, `clubs/${clubId}/members/${memberId}`);
    await updateDoc(memberRef, {
      certificat_medical_validite: Timestamp.fromDate(validUntil),
      has_pending_medical: false // Clear the pending flag
    });
  }

  /**
   * Reject a medical certificate
   * Sets the status to 'rejected' with a reason
   */
  static async rejectCertification(
    clubId: string,
    memberId: string,
    certId: string,
    reason: string,
    reviewerId: string,
    reviewerName: string
  ): Promise<void> {
    const certRef = doc(db, `clubs/${clubId}/members/${memberId}/medical_certificates/${certId}`);

    await updateDoc(certRef, {
      status: 'rejected' as CertificateStatus,
      rejection_reason: reason,
      reviewed_by: reviewerId,
      reviewed_by_nom: reviewerName,
      reviewed_at: Timestamp.fromDate(new Date())
    });

    // Check if there are any remaining pending certificates
    const certs = await this.getMemberCertifications(clubId, memberId);
    const hasPending = certs.some(c => c.status === 'pending');

    // Update the member's pending flag
    const memberRef = doc(db, `clubs/${clubId}/members/${memberId}`);
    await updateDoc(memberRef, {
      has_pending_medical: hasPending
    });
  }

  /**
   * Delete a medical certificate (and its file from storage)
   */
  static async deleteCertification(
    clubId: string,
    memberId: string,
    certId: string,
    documentUrl: string
  ): Promise<void> {
    // Delete from Storage first
    try {
      const storageRef = ref(storage, documentUrl);
      await deleteObject(storageRef);
    } catch (e) {
      // File may not exist, continue with Firestore cleanup
      logger.warn('File not found in storage:', e);
    }

    // Delete from Firestore
    const certRef = doc(db, `clubs/${clubId}/members/${memberId}/medical_certificates/${certId}`);
    await deleteDoc(certRef);

    // Check if there are any remaining pending certificates
    const certs = await this.getMemberCertifications(clubId, memberId);
    const hasPending = certs.some(c => c.status === 'pending');

    // Update the member's pending flag
    const memberRef = doc(db, `clubs/${clubId}/members/${memberId}`);
    await updateDoc(memberRef, {
      has_pending_medical: hasPending
    });
  }

  /**
   * Get the current (most relevant) certificate for a member
   * Priority: approved (by valid_until) > pending (by uploaded_at) > rejected
   */
  static getCurrentCertificate(
    certs: MobileMedicalCertificate[]
  ): MobileMedicalCertificate | null {
    if (certs.length === 0) return null;

    // Priority 1: Approved certificates (most recent by valid_until)
    const approved = certs
      .filter(c => c.status === 'approved')
      .sort((a, b) => {
        const aDate = a.valid_until?.toDate?.() || new Date(0);
        const bDate = b.valid_until?.toDate?.() || new Date(0);
        return bDate.getTime() - aDate.getTime();
      });

    if (approved.length > 0) return approved[0];

    // Priority 2: Pending certificates (most recent by uploaded_at)
    const pending = certs
      .filter(c => c.status === 'pending')
      .sort((a, b) => {
        const aDate = a.uploaded_at?.toDate?.() || new Date(0);
        const bDate = b.uploaded_at?.toDate?.() || new Date(0);
        return bDate.getTime() - aDate.getTime();
      });

    if (pending.length > 0) return pending[0];

    // Priority 3: Rejected (most recent)
    return certs[0]; // Already sorted by uploaded_at desc
  }

  /**
   * Check if a certificate is valid (approved and not expired)
   */
  static isValid(cert: MobileMedicalCertificate): boolean {
    if (cert.status !== 'approved' || !cert.valid_until) return false;
    const validUntil = cert.valid_until.toDate?.() || new Date(cert.valid_until as any);
    return validUntil > new Date();
  }

  /**
   * Check if a certificate is expiring soon (within 30 days)
   */
  static isExpiringSoon(cert: MobileMedicalCertificate): boolean {
    if (cert.status !== 'approved' || !cert.valid_until) return false;
    const validUntil = cert.valid_until.toDate?.() || new Date(cert.valid_until as any);
    const thirtyDaysFromNow = new Date();
    thirtyDaysFromNow.setDate(thirtyDaysFromNow.getDate() + 30);
    return validUntil > new Date() && validUntil <= thirtyDaysFromNow;
  }

  /**
   * Get the default validity date (31 January of next year)
   */
  static getDefaultValidityDate(): Date {
    const now = new Date();
    const nextYear = now.getFullYear() + 1;
    return new Date(nextYear, 0, 31); // January 31 of next year
  }

  /**
   * Sync the has_pending_medical flag on the member document
   * This fixes the flag for certificates uploaded before the mobile app was updated
   */
  static async syncPendingFlag(
    clubId: string,
    memberId: string,
    certs: MobileMedicalCertificate[]
  ): Promise<void> {
    const hasPending = certs.some(c => c.status === 'pending');
    const memberRef = doc(db, `clubs/${clubId}/members/${memberId}`);
    await updateDoc(memberRef, { has_pending_medical: hasPending });
  }

  /**
   * Update the display name of a mobile certificate
   */
  static async updateCertificateName(
    clubId: string,
    memberId: string,
    certId: string,
    newName: string
  ): Promise<void> {
    const certRef = doc(db, `clubs/${clubId}/members/${memberId}/medical_certificates/${certId}`);
    await updateDoc(certRef, { file_name: newName });
  }

  /**
   * Update the validity date of a mobile certificate
   * Also updates the member's main validity date if this is the most recent
   */
  static async updateCertificateValidityDate(
    clubId: string,
    memberId: string,
    certId: string,
    newValidUntil: Date
  ): Promise<void> {
    const certRef = doc(db, `clubs/${clubId}/members/${memberId}/medical_certificates/${certId}`);
    await updateDoc(certRef, { valid_until: Timestamp.fromDate(newValidUntil) });

    // Check if we need to update the member's main validity date
    const certs = await this.getMemberCertifications(clubId, memberId);
    const approvedCerts = certs
      .filter(c => c.status === 'approved' && c.valid_until)
      .sort((a, b) => {
        const aDate = a.valid_until?.toDate?.() || new Date(0);
        const bDate = b.valid_until?.toDate?.() || new Date(0);
        return bDate.getTime() - aDate.getTime();
      });

    if (approvedCerts.length > 0) {
      const latestDate = approvedCerts[0].valid_until?.toDate?.() || new Date(approvedCerts[0].valid_until as any);
      const memberRef = doc(db, `clubs/${clubId}/members/${memberId}`);
      await updateDoc(memberRef, { certificat_medical_validite: Timestamp.fromDate(latestDate) });
    }
  }
}

File: lib/services/medical_certification_service.dart
import 'dart:io';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_storage/firebase_storage.dart';
import 'package:flutter/foundation.dart';
import '../models/medical_certification.dart';

/// Service de gestion des certificats médicaux
class MedicalCertificationService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;
  final FirebaseStorage _storage = FirebaseStorage.instance;

  /// Récupérer le certificat médical actuel (le plus récent approuvé ou en attente)
  Future<MedicalCertification?> getCurrentCertification(
    String clubId,
    String userId,
  ) async {
    try {
      // D'abord chercher un certificat approuvé et valide
      final approvedQuery = await _firestore
          .collection('clubs/$clubId/members/$userId/medical_certificates')
          .where('status', isEqualTo: 'approved')
          .orderBy('valid_until', descending: true)
          .limit(1)
          .get();

      if (approvedQuery.docs.isNotEmpty) {
        final cert = MedicalCertification.fromFirestore(approvedQuery.docs.first);
        debugPrint('✅ Certificat approuvé trouvé: ${cert.id}');
        return cert;
      }

      // Sinon chercher un certificat en attente
      final pendingQuery = await _firestore
          .collection('clubs/$clubId/members/$userId/medical_certificates')
          .where('status', isEqualTo: 'pending')
          .orderBy('uploaded_at', descending: true)
          .limit(1)
          .get();

      if (pendingQuery.docs.isNotEmpty) {
        final cert = MedicalCertification.fromFirestore(pendingQuery.docs.first);
        debugPrint('⏳ Certificat en attente trouvé: ${cert.id}');
        return cert;
      }

      debugPrint('ℹ️ Aucun certificat médical trouvé pour $userId');
      return null;
    } catch (e) {
      debugPrint('❌ Erreur récupération certificat: $e');
      return null;
    }
  }

  /// Stream du certificat médical actuel (temps réel)
  /// Priorité: certificat approuvé > certificat en attente
  /// (même logique que getCurrentCertification)
  Stream<MedicalCertification?> watchCurrentCertification(
    String clubId,
    String userId,
  ) {
    // Watch all certificates and apply priority logic locally
    return _firestore
        .collection('clubs/$clubId/members/$userId/medical_certificates')
        .orderBy('uploaded_at', descending: true)
        .snapshots()
        .map((snapshot) {
      if (snapshot.docs.isEmpty) return null;

      final certs = snapshot.docs
          .map((doc) => MedicalCertification.fromFirestore(doc))
          .toList();

      // Priority 1: Find approved certificate (most recent by valid_until)
      final approvedCerts = certs
          .where((c) => c.status == CertificateStatus.approved)
          .toList();
      if (approvedCerts.isNotEmpty) {
        // Sort by valid_until descending
        approvedCerts.sort((a, b) {
          final aDate = a.validUntil ?? DateTime(1900);
          final bDate = b.validUntil ?? DateTime(1900);
          return bDate.compareTo(aDate);
        });
        return approvedCerts.first;
      }

      // Priority 2: Find pending certificate (most recent by uploaded_at)
      final pendingCerts = certs
          .where((c) => c.status == CertificateStatus.pending)
          .toList();
      if (pendingCerts.isNotEmpty) {
        // Already sorted by uploaded_at descending from query
        return pendingCerts.first;
      }

      // Priority 3: Return most recent rejected (for display)
      return certs.first;
    });
  }

  /// Récupérer l'historique des certificats
  Future<List<MedicalCertification>> getCertificationHistory(
    String clubId,
    String userId,
  ) async {
    try {
      final snapshot = await _firestore
          .collection('clubs/$clubId/members/$userId/medical_certificates')
          .orderBy('uploaded_at', descending: true)
          .get();

      final certs = snapshot.docs
          .map((doc) => MedicalCertification.fromFirestore(doc))
          .toList();

      debugPrint('✅ ${certs.length} certificat(s) trouvé(s)');
      return certs;
    } catch (e) {
      debugPrint('❌ Erreur récupération historique: $e');
      return [];
    }
  }

  /// Détecte le type réel du document à partir de ses octets d'en-tête.
  /// Retourne 'pdf' si le fichier commence par la signature `%PDF-`,
  /// sinon 'image'. En cas d'erreur de lecture, on retombe sur le type
  /// annoncé par l'appelant (normalisé à 'image' si inconnu).
  Future<String> _detectDocumentType(File file, String fallbackType) async {
    try {
      final raf = await file.open();
      final header = await raf.read(5);
      await raf.close();
      final isPdf = header.length >= 5 &&
          header[0] == 0x25 && // %
          header[1] == 0x50 && // P
          header[2] == 0x44 && // D
          header[3] == 0x46 && // F
          header[4] == 0x2D; // -
      if (isPdf) return 'pdf';
      // Le contenu n'est pas un PDF : c'est forcément une image.
      return 'image';
    } catch (e) {
      debugPrint('⚠️ Détection type fichier échouée, fallback "$fallbackType": $e');
      return fallbackType == 'pdf' ? 'pdf' : 'image';
    }
  }

  /// Uploader un nouveau certificat médical
  Future<MedicalCertification> uploadCertification({
    required String clubId,
    required String userId,
    required File file,
    required String documentType, // 'image' ou 'pdf'
    String? fileName,
  }) async {
    try {
      debugPrint('📤 Upload certificat médical pour $userId...');

      // 0. Sécurité: déterminer le VRAI type via les octets d'en-tête.
      // L'extension transmise par l'appelant (file_picker / scanner) n'est pas
      // fiable — notamment nulle sur iOS — ce qui faisait jadis téléverser des
      // PDF étiquetés image/jpeg, affichés cassés dans CalyCompta.
      final effectiveType = await _detectDocumentType(file, documentType);

      // 1. Générer un nom de fichier unique
      final timestamp = DateTime.now().millisecondsSinceEpoch;
      final extension = effectiveType == 'pdf' ? 'pdf' : 'jpg';
      final storageName = 'certificat_$timestamp.$extension';

      // 2. Chemin dans Storage
      final ref = _storage
          .ref()
          .child('clubs/$clubId/members/$userId/medical_certificates/$storageName');

      // 3. Upload avec metadata
      final contentType =
          effectiveType == 'pdf' ? 'application/pdf' : 'image/jpeg';
      final uploadTask = await ref.putFile(
        file,
        SettableMetadata(
          contentType: contentType,
          customMetadata: {
            'uploadedBy': userId,
            'uploadedAt': DateTime.now().toIso8601String(),
            'originalFileName': fileName ?? storageName,
          },
        ),
      );

      // 4. Récupérer l'URL de téléchargement
      final downloadUrl = await uploadTask.ref.getDownloadURL();
      debugPrint('✅ Fichier uploadé: $downloadUrl');

      // 5. Créer le document Firestore
      final now = DateTime.now();
      final docRef = _firestore
          .collection('clubs/$clubId/members/$userId/medical_certificates')
          .doc();

      final certification = MedicalCertification(
        id: docRef.id,
        memberId: userId,
        documentUrl: downloadUrl,
        documentType: effectiveType,
        fileName: fileName,
        uploadedAt: now,
        status: CertificateStatus.pending,
      );

      await docRef.set(certification.toFirestore());
      debugPrint('✅ Certificat créé: ${docRef.id}');

      // 6. Update member's has_pending_medical flag
      final memberRef = _firestore.doc('clubs/$clubId/members/$userId');
      await memberRef.update({'has_pending_medical': true});
      debugPrint('✅ Flag has_pending_medical mis à jour');

      return certification;
    } catch (e) {
      debugPrint('❌ Erreur upload certificat: $e');
      rethrow;
    }
  }

  /// Supprimer un certificat médical
  Future<void> deleteCertification(
    String clubId,
    String userId,
    String certId,
  ) async {
    try {
      // 1. Récupérer le document pour avoir l'URL du fichier
      final docRef = _firestore
          .collection('clubs/$clubId/members/$userId/medical_certificates')
          .doc(certId);

      final doc = await docRef.get();
      if (!doc.exists) {
        debugPrint('⚠️ Certificat $certId non trouvé');
        return;
      }

      final cert = MedicalCertification.fromFirestore(doc);

      // 2. Supprimer le fichier de Storage
      try {
        final ref = _storage.refFromURL(cert.documentUrl);
        await ref.delete();
        debugPrint('✅ Fichier supprimé de Storage');
      } catch (e) {
        debugPrint('⚠️ Erreur suppression fichier Storage: $e');
      }

      // 3. Supprimer le document Firestore
      await docRef.delete();
      debugPrint('✅ Certificat $certId supprimé');
    } catch (e) {
      debugPrint('❌ Erreur suppression certificat: $e');
      rethrow;
    }
  }

  /// Vérifier si le membre a un certificat valide
  Future<bool> hasValidCertification(String clubId, String userId) async {
    final cert = await getCurrentCertification(clubId, userId);
    return cert?.canPresent ?? false;
  }
}
